scalar: rework `needs_git_repo`

dscho opened the pull request:
We cannot actually set up the gitdir before parsing the options because we want to support the optional enlistment path parameter. Meaning: the `cmd_*()` function must be allowed to parse the options _before_ we know whether that optional enlistment path parameter was provided or not.

While at it, we change the paradigm by always `chdir()`-ing into the worktree. That allows us to simplify the code rather dramatically, as we no longer have to pass around the `dir` parameter, and we also need not care to specify the exact path to the local repository config anymore.

dscho committed:
fixup! scalar register: set recommended config settings

We need to enforce the distinction between an enlistment and a worktree.
A Scalar enlistment is the top-level directory in which the `src/`
directory lives, which is a Git worktree.

The idea is to encourage as few untracked files within a worktree as
possible.

The `scalar register` command expects an optional enlistment path, not
a worktree path.

Signed-off-by: Johannes Schindelin <johannes.schindelin@gmx.de>

dscho committed:
fixup! scalar register/unregister: start/stop maintenance on repository

Let's handle the optional `<enlistment>` parameter correctly (it need
not be the worktree).

Signed-off-by: Johannes Schindelin <johannes.schindelin@gmx.de>

dscho committed:
fixup! scalar: implement the `run` command

This teaches the `run` command about the optional enlistment parameter.

Signed-off-by: Johannes Schindelin <johannes.schindelin@gmx.de>

dscho committed:
fixup! scalar: add the `cache-server` command

The optional parameter refers to the enlistment, not the worktree. The
difference: in a Scalar enlistment <dir>, the worktree is always in
<dir>/src/.

Signed-off-by: Johannes Schindelin <johannes.schindelin@gmx.de>

dscho committed:
fixup! Implement `scalar diagnose`

This teaches `scalar diagnose` about that optional enlistment parameter.

Signed-off-by: Johannes Schindelin <johannes.schindelin@gmx.de>

dscho committed:
fixup! Start porting `scalar.exe` to C

We cannot actually discover the Git directory before handing off to the
`cmd_*()` functions because they receive optional _enlistment_ path
parameters (and without parsing the options, it is impossible to say
whether this parameter has been provided or not).

Signed-off-by: Johannes Schindelin <johannes.schindelin@gmx.de>

dscho committed:
fixup! scalar: implement the `clone` subcommand

If we `chdir()` into the freshly-initialized worktree, and _then_
discove the just-created gitdir, we can avoid a whole slew of complex
code. So let's do that.

Signed-off-by: Johannes Schindelin <johannes.schindelin@gmx.de>

dscho committed:
fixup! scalar register: set recommended config settings

Now that there are no callers left that pass a non-NULL `dir` parameter,
we can remove that parameter.

Signed-off-by: Johannes Schindelin <johannes.schindelin@gmx.de>

dscho committed:
fixup! scalar register/unregister: start/stop maintenance on repository

The `unregister_dir()` function is only called with `NULL` as a
parameter. Let's remove that parameter altogether.

Signed-off-by: Johannes Schindelin <johannes.schindelin@gmx.de>

dscho committed:
fixup! scalar unregister: stop FSMonitor daemon

The only caller of this function passes `NULL` as `dir`, therefore this
parameter is not even needed.

Signed-off-by: Johannes Schindelin <johannes.schindelin@gmx.de>

dscho committed:
fixup! scalar: implement 'scalar list'

The only callers of `add_or_remove_enlistment()` passed `NULL` as `dir`
parameter, therefore it can be removed altogether.

Signed-off-by: Johannes Schindelin <johannes.schindelin@gmx.de>

dscho committed:
fixup! scalar register/unregister: start/stop maintenance on repository

The `dir` parameter of `toggle_maintenance()` is no longer needed.

Signed-off-by: Johannes Schindelin <johannes.schindelin@gmx.de>

dscho committed:
fixup! scalar register: set recommended config settings

The parameter of `set_recommended_config()` is no longer needed: we
always call this function _after_ discovering the current gitdir.

Signed-off-by: Johannes Schindelin <johannes.schindelin@gmx.de>

dscho committed:
fixup! scalar: implement the `clone` subcommand

We only call `set_config()` after discovering the gitdir now, which
means we do not need to specify the path to the config file explicitly
anymore (which is nice because it simplifies the code).

Signed-off-by: Johannes Schindelin <johannes.schindelin@gmx.de>

dscho committed:
fixup! scalar register/unregister: start/stop maintenance on repository

Just like all the other functions, `run_git()` no longer needs the `dir`
parameter.

Signed-off-by: Johannes Schindelin <johannes.schindelin@gmx.de>

dscho committed:
fixup! scalar clone: support GVFS-enabled remote repositories

The `dir` parameter of the `supports_gvfs_protocol()` function turns out
to be unnecessary.

Signed-off-by: Johannes Schindelin <johannes.schindelin@gmx.de>

mjcheetham reviewed (approved):
This makes the code of the commands much nicer and simpler, for the payment of a small bit of boilerplate.. nice!

Did we want to codify (in docs) the term "enlistment" vs "worktree"?

> Enlistment is the directory containing the Scalar worktree named "src".

(or similar)?

derrickstolee commented on contrib/scalar/t/t9099-scalar.sh:
```diff
@@ -59,19 +59,15 @@ test_expect_success 'scalar clone' '
-		unzip -p "$zip_path" diagnostics.log >out &&
-		test_file_not_empty out
-	)
+	scalar diagnose cloned >out &&
```
Nice test.

dscho replied:
Isn't it? ;-)

derrickstolee reviewed (approved):
> This makes the code of the commands much nicer and simpler, for the payment of a small bit of boilerplate.. nice!

I agree.
 
> Did we want to codify (in docs) the term "enlistment" vs "worktree"?
> 
> > Enlistment is the directory containing the Scalar worktree named "src".
> 
> (or similar)?

We should take a round of updating the docs after the functionality is hammered out and the `microsoft/scalar` functional tests are passing.

dscho committed:
fixup! scalar: do initialize `gvfs.sharedCache`

We need to be careful not to use `hash_to_hex()`: now that we call
`setup_git_directory()`, this will correctly assume SHA-256 when running
under `GIT_TEST_DEFAULT_HASH=sha256`, but we _really_ want to use SHA-1
here.

Signed-off-by: Johannes Schindelin <johannes.schindelin@gmx.de>

dscho commented:
> Did we want to codify (in docs) the term "enlistment" vs "worktree"?

Right, I forgot to update the documentation!

dscho committed:
fixup! scalar: start documenting the command

Let's document the concept of an "enlistment" right off the bat.

Signed-off-by: Johannes Schindelin <johannes.schindelin@gmx.de>

dscho committed:
fixup! scalar: document the `clone` subcommand

Let's use the term "enlistment" instead of "directory".

Signed-off-by: Johannes Schindelin <johannes.schindelin@gmx.de>

dscho committed:
fixup! scalar: document `list`, `register` and `unregister`

Let's use the term "enlistment" instead of "path".

Signed-off-by: Johannes Schindelin <johannes.schindelin@gmx.de>

dscho committed:
fixup! scalar: document the remaining subcommands

Let's document the optional `<enlistment>` parameter.

Signed-off-by: Johannes Schindelin <johannes.schindelin@gmx.de>

dscho committed:
fixup! scalar: add the `cache-server` command

Let's consistently document the optional `<enlistment>` parameter.

Signed-off-by: Johannes Schindelin <johannes.schindelin@gmx.de>

dscho commented:
> 
> 
> > Did we want to codify (in docs) the term "enlistment" vs "worktree"?
> 
> Right, I forgot to update the documentation!

There you go: https://github.com/microsoft/git/compare/5d5bf2d99ce49a6dd79c681135cada5334757c7c...5dcb61158fe960f7eb3057f13a82af6ef8c6856f
